File: controller/discovery/diagnosis.go
// Copyright 2026 ZyvorAI Labs Private Limited
// SPDX-License-Identifier: Apache-2.0

package discovery

import (
	"context"
	"fmt"
	"net/url"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/labels"

	"github.com/zyvorai/hermes/controller/model"
)

func (w *Watcher) Diagnose(ctx context.Context, app model.App) model.AppDiagnosis {
	slug := app.CanonicalSlug
	if slug == "" {
		slug = app.Slug
	}
	chain := []model.DiagnosisChainNode{
		{ID: "user", Label: "User"},
		{ID: "identity", Label: "Zeus Identity"},
		{ID: "gateway", Label: "Hermes Gateway"},
		{ID: "approute", Label: "AppRoute: " + slug},
		{ID: "namespace", Label: "Namespace: " + app.Namespace},
	}

	svcLabel := fmt.Sprintf("%s:%d", app.Backend.Name, app.Backend.Port)
	svcStatus := app.Status
	if app.ReadyCount == 0 {
		svcStatus = model.StatusBroken
	}
	chain = append(chain, model.DiagnosisChainNode{
		ID: "service", Label: "Service " + svcLabel, Status: svcStatus,
	})

	problem := ""
	cause := ""
	if app.Status != model.StatusHealthy {
		problem = app.StatusMsg
		if problem == "" {
			switch app.Status {
			case model.StatusBroken:
				problem = "Backend unavailable"
			case model.StatusDegraded:
				problem = "Backend degraded"
			default:
				problem = "Health check failed"
			}
		}
	}

	podName, podStatus := w.lookupPodHint(ctx, app)
	if podName != "" {
		chain = append(chain, model.DiagnosisChainNode{
			ID: "pod", Label: "Pod " + podName, Status: podStatus,
		})
		if cause == "" && podStatus != "" {
			cause = fmt.Sprintf("Pod %s is %s", podName, podStatus)
		}
	} else if app.ReadyCount == 0 {
		cause = "Service exists but has no ready endpoints"
	}

	ns := url.QueryEscape(app.Namespace)
	actions := []model.SuggestedAction{
		{Label: "Open Kubernetes workloads", Href: fmt.Sprintf("/k8s/workloads?ns=%s", ns)},
	}
	if podName != "" {
		actions = append(actions, model.SuggestedAction{
			Label: "View pod logs",
			Href:  fmt.Sprintf("/k8s/workloads?ns=%s", ns),
		})
	}

	return model.AppDiagnosis{
		AppID:            app.ID,
		RoutePath:        app.RoutePath,
		PublicURL:        app.PublicURL,
		Backend:          app.Backend,
		Problem:          problem,
		Cause:            cause,
		Chain:            chain,
		SuggestedActions: actions,
	}
}

func (w *Watcher) lookupPodHint(ctx context.Context, app model.App) (string, string) {
	svc, err := w.client.CoreV1().Services(app.Namespace).Get(ctx, app.Backend.Name, metav1.GetOptions{})
	if err != nil {
		return "", ""
	}
	selector := labels.Set(svc.Spec.Selector).String()
	if selector == "" {
		return "", ""
	}
	pods, err := w.client.CoreV1().Pods(app.Namespace).List(ctx, metav1.ListOptions{
		LabelSelector: selector,
	})
	if err != nil || len(pods.Items) == 0 {
		return "", ""
	}
	pod := pods.Items[0]
	status := string(pod.Status.Phase)
	for _, cs := range pod.Status.ContainerStatuses {
		if cs.State.Waiting != nil && cs.State.Waiting.Reason != "" {
			status = cs.State.Waiting.Reason
			break
		}
		if cs.State.Terminated != nil && cs.State.Terminated.Reason != "" {
			status = cs.State.Terminated.Reason
			break
		}
	}
	return pod.Name, status
}

File: crates/hermes-core/src/diagnosis.rs
// Copyright 2026 ZyvorAI Labs Private Limited
// SPDX-License-Identifier: Apache-2.0

use serde::{Deserialize, Serialize};

use crate::{App, Backend};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DiagnosisChainNode {
    pub id: String,
    pub label: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SuggestedAction {
    pub label: String,
    pub href: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AppDiagnosis {
    pub app_id: String,
    pub route_path: String,
    pub public_url: String,
    pub backend: Backend,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub problem: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub cause: String,
    pub chain: Vec<DiagnosisChainNode>,
    pub suggested_actions: Vec<SuggestedAction>,
}

pub fn build_diagnosis(app: &App) -> AppDiagnosis {
    let slug = if app.canonical_slug.is_empty() {
        app.slug.as_str()
    } else {
        app.canonical_slug.as_str()
    };
    let mut chain = vec![
        DiagnosisChainNode {
            id: "user".into(),
            label: "User".into(),
            status: String::new(),
        },
        DiagnosisChainNode {
            id: "identity".into(),
            label: "Zeus Identity".into(),
            status: String::new(),
        },
        DiagnosisChainNode {
            id: "gateway".into(),
            label: "Hermes Gateway".into(),
            status: String::new(),
        },
        DiagnosisChainNode {
            id: "approute".into(),
            label: format!("AppRoute: {slug}"),
            status: String::new(),
        },
        DiagnosisChainNode {
            id: "namespace".into(),
            label: format!("Namespace: {}", app.namespace),
            status: String::new(),
        },
    ];

    let svc_label = format!("{}:{}", app.backend.name, app.backend.port);
    let mut svc_status = app.status.clone();
    if app.ready_endpoints == 0 {
        svc_status = "broken".into();
    }
    chain.push(DiagnosisChainNode {
        id: "service".into(),
        label: format!("Service {svc_label}"),
        status: svc_status,
    });

    let mut problem = String::new();
    let mut cause = String::new();
    if app.status != "healthy" {
        problem = app.status_message.clone();
        if problem.is_empty() {
            problem = match app.status.as_str() {
                "broken" => "Backend unavailable".into(),
                "degraded" => "Backend degraded".into(),
                _ => "Health check failed".into(),
            };
        }
        if app.ready_endpoints == 0 {
            cause = "Service exists but has no ready endpoints".into();
        }
    }

    let ns = app.namespace.replace(' ', "%20");
    let mut suggested_actions = vec![
        SuggestedAction {
            label: "View in cluster catalog".into(),
            href: format!("/cluster?ns={ns}"),
        },
        SuggestedAction {
            label: "Open dependency graph".into(),
            href: "/graph".into(),
        },
    ];
    if app.ready_endpoints == 0 || app.status != "healthy" {
        suggested_actions.push(SuggestedAction {
            label: "Copy kubectl pod check".into(),
            href: format!(
                "#copy:kubectl get pods -n {} -l app.kubernetes.io/name={}",
                app.namespace, app.backend.name
            ),
        });
    }
    suggested_actions.push(SuggestedAction {
        label: "Open Kubernetes workloads".into(),
        href: format!("/k8s/workloads?ns={ns}"),
    });

    if !app.meta.ingress_hosts.is_empty() {
        let host = app.meta.ingress_hosts[0].clone();
        let href = if host.starts_with("http") {
            host.clone()
        } else {
            format!("https://{host}")
        };
        suggested_actions.insert(
            0,
            SuggestedAction {
                label: format!("Open ingress: {host}"),
                href,
            },
        );
    }

    AppDiagnosis {
        app_id: app.id.clone(),
        route_path: app.route_path.clone(),
        public_url: app.public_url.clone(),
        backend: app.backend.clone(),
        problem,
        cause,
        chain,
        suggested_actions,
    }
}
